engine/access/rpc/backend/backend_events.go

```diff
@@ -100,27 +116,34 @@ func (b *backendEvents) getBlockEventsFromExecutionNode(
 
 // verifyAndConvertToAccessEvents converts execution node api result to access node api result, and verifies that the results contains
 // results from each block that was requested
-func verifyAndConvertToAccessEvents(execEvents []*execproto.GetEventsForBlockIDsResponse_Result, requestedBlockIDs []flow.Identifier) ([]flow.BlockEvents, error) {
-	if len(execEvents) != len(requestedBlockIDs) {
+func verifyAndConvertToAccessEvents(execEvents []*execproto.GetEventsForBlockIDsResponse_Result, requestedBlockHeaders []*flow.Header) ([]flow.BlockEvents, error) {
+	if len(execEvents) != len(requestedBlockHeaders) {
 		return nil, errors.New("number of results does not match number of blocks requested")
 	}
 
-	blockIDSet := map[string]bool{}
-	for _, blockID := range requestedBlockIDs {
-		blockIDSet[blockID.String()] = true
+	reqestedBlockHeaderSet := map[string]*flow.Header{}
+	for _, header := range requestedBlockHeaders {
+		reqestedBlockHeaderSet[header.ID().String()] = header
 	}
 
 	results := make([]flow.BlockEvents, len(execEvents))
 
 	for i, result := range execEvents {
-		if !blockIDSet[hex.EncodeToString(result.GetBlockId())] {
+		header, expected := reqestedBlockHeaderSet[hex.EncodeToString(result.GetBlockId())]
+		if !expected {
 			return nil, fmt.Errorf("unexpected blockID from exe node %x", result.GetBlockId())
 		}
+		if result.GetBlockHeight() != header.Height {
+			return nil, fmt.Errorf("unexpected block height %x for block %x from exe node",
```

Comment:
```suggestion
			return nil, fmt.Errorf("unexpected block height %d for block %x from exe node",
```

`%s` is for printing `[]byte` as a hex string